keps/115-Subgroup-support/README.md

```diff
@@ -0,0 +1,227 @@
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+To avoid the changes from affecting initial functionality and to not set 
+unnecessary labels, there will be no default behavior if `SubGroupSize` is not 
+set
+
+## Design Details
+
+The overall goal of subgroup scheduling
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+### LeaderWorkerSet API
+We extend the LeaderWorkerSet API to introduce a new field: subGroupSize to opt in and set the number of pods that each subgroup will contain. Current behavior is kept if not set. 
+
+```
+type LeaderWorkerTemplate struct {
+	 // Number of pods per subgroup. This value is immutable,
+   // and must not be greater than LeaderWorkerSet.Spec.Size.
+	SubGroupSize *int32 'json:"subGroupSize,omitempty'
+} 
+```
+
+### Exclusive Topology Support
+LeaderWorkerSet can guarantee that the leader and the workers are placed in the same topology if the `leaderworkerset.sigs.k8s.io/exclusive-topology` annotation is set. Similarly, we will support that the pods within the same subgroup will be placed in the same topology with a new annotation `leaderworkerset.sigs.k8s.io/subgroup-exclusive-topology`, so that the new changes can support up to two levels of pod affinity. 
+
+The overall workflow will look like this 
+
+![kep-115](https://github.com/kubernetes-sigs/lws/assets/86417275/ff9fc93d-c738-4c09-abc8-50a7b16d49df)
+
+Suppose we have an LWS deployment with 2 replicas, subGroupSize 4, size 8, and the following annotations: 
+- `leaderworkerset.sigs.k8s.io/exclusive-topology: topology-1` 
+- `leaderworkerset.sigs.k8s.io/subgroup-exclusive-topology: topology-2`
+
+The leader pod will first be scheduled on a placement group. Once a worker pod is created, it will follow the leader to the placement group. 
+Afterward, it will be scheduled into a nodepool that has other pods with the same subgroup index. So, the placement will look something like this
+
+![kep-115 scheduling](https://github.com/kubernetes-sigs/lws/assets/86417275/33c49e9f-84da-4ac9-8e2b-973b532a0756)
+
+### Implementation
+- Two new annotations will be added
+  - `leaderworkerset.sigs.k8s.io/subgroup-exclusive-topology`
+  - `leaderworkerset.sigs.k8s.io/subgroup-size` 
+- Three new labels will be added,
+  - `leaderworkerset.sigs.k8s.io/subgroup-index = worker-index/subGroupSize`
+    - Tracks which subgroup the pod is part of 
```

Comment:
I don't think we need this for now, mostly because I am not sure about the semantics we want to have for it if the leader is not considered an actual "worker" (the case where the first subgroup has an extra pod, the leader).

Reply:
removed subgroup-worker-index